sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AsyncOffsetSeqLog.scala

```diff
@@ -0,0 +1,185 @@
+import org.apache.spark.util.{Clock, SystemClock}
+
+/**
+ * Used to write entries to the offset log asynchronously
+ */
+class AsyncOffsetSeqLog(
+    sparkSession: SparkSession,
+    path: String,
+    executorService: ThreadPoolExecutor,
+    offsetCommitIntervalMs: Long,
+    clock: Clock = new SystemClock())
+    extends OffsetSeqLog(sparkSession, path) {
+
+  // the cache needs to be enabled because we may not be persisting every entry to durable storage
+  // entries not persisted to durable storage will just be stored in memory for faster lookups
+  assert(metadataCacheEnabled == true)
+
+  // A map of the current pending offset writes. Key -> batch Id, Value -> CompletableFuture
+  // Used to determine if a commit log entry for this batch also needs to be persisted to storage
+  private val pendingOffsetWrites = new ConcurrentHashMap[Long, CompletableFuture[Long]]()
+
+  // Keeps track the last time a commit was issued. Used for issuing commits to storage at
+  // the configured intervals
+  private val lastCommitIssuedTimestampMs: AtomicLong = new AtomicLong(-1)
+
+  // A queue of batches written to storage.  Used to keep track when to purge old batches
+  val writtenToDurableStorage =
+    new ConcurrentLinkedDeque[Long](listBatchesOnDisk.toList.asJavaCollection)
+
+  /**
+   * Get a async offset write by batch id.  To check if a corresponding commit log entry
+   * needs to be written to durable storage as well
+   * @param batchId
+   * @return a option to indicate whether a async offset write was issued for the batch with id
+   */
+  def getAsyncOffsetWrite(batchId: Long): Option[CompletableFuture[Long]] = {
+    Option(pendingOffsetWrites.get(batchId))
+  }
+
+  /**
+   * Remove the async offset write when we don't need to keep track of it anymore
+   * @param batchId
+   */
+  def removeAsyncOffsetWrite(batchId: Long): Unit = {
+    pendingOffsetWrites.remove(batchId)
+  }
+
+  /**
+   * Writes a new batch to the offset log asynchronously
+   * @param batchId id of batch to write
+   * @param metadata metadata of batch to write
+   * @return a CompeletableFuture that contains the batch id.  The future is completed when
+   *         the async write of the batch is completed.  Future may also be completed exceptionally
+   *         to indicate some write error.
+   */
+  def addAsync(batchId: Long, metadata: OffsetSeq): CompletableFuture[(Long, Boolean)] = {
+    require(metadata != null, "'null' metadata cannot written to a metadata log")
+
+    def issueAsyncWrite(batchId: Long): CompletableFuture[Long] = {
+      lastCommitIssuedTimestampMs.set(clock.getTimeMillis())
+      val future: CompletableFuture[Long] = addNewBatchByStreamAsync(batchId) { output =>
+        serialize(metadata, output)
+      }.thenApply((ret: Boolean) => {
+        if (ret) {
+          batchId
+        } else {
+          throw new IllegalStateException(
+            s"Concurrent update to the log. Multiple streaming jobs detected for $batchId"
```

Comment:
Shall we be more specific for "where"? offset? commit?

Reply:
This is the message in existing implementation. We don't really differentiate which log. However, you should know which log it is based on the stack trace.